### ISSUE TITLE
Just bicycle made

### ISSUE THREAD

**Issue by elnazbanani** [HARPLab/DReyeVR#98]
Hi. Although I call the vehicle with attribute ('number_of_wheels')) == 4, it just provide bicycle. this is my code. could you help me?

import time
import carla
#import numpy as np
#from DReyeVR_utils import find_ego_vehicle

from carla import VehicleLightState as vls

import argparse
import logging

from numpy import random

#---------------get actor----------------------------
def get_actor_blueprints(world, filter, generation):
    bps = world.get_blueprint_library().filter(filter)

    if generation.lower() == "all":
        return bps

    # If the filter returns only one bp, we assume that this one needed
    # and therefore, we ignore the generation
    if len(bps) == 1:
        return bps

    try:
        int_generation = int(generation)
        # Check if generation is in available generations
        if int_generation in [1, 2]:
            bps = [x for x in bps if int(x.get_attribute('generation')) == int_generation]
            return bps
        else:
            print("   Warning! Actor Generation is not valid. No actor will be spawned.")
            return []
    except:
        print("   Warning! Actor Generation is not valid. No actor will be spawned.")
        return []
#---------------------------------------------------------------
def main():
    argparser = argparse.ArgumentParser(
        description=__doc__)
    argparser.add_argument(
        '--host',
        metavar='H',
        default='127.0.0.1',
        help='IP of the host server (default: 127.0.0.1)')
    argparser.add_argument(
        '-p', '--port',
        metavar='P',
        default=2000,
        type=int,
        help='TCP port to listen to (default: 2000)')
    argparser.add_argument(
        '-n', '--number-of-vehicles',
        metavar='N',
        default=50,
        type=int,
        help='Number of vehicles (default: 30)')
    argparser.add_argument(
        '--safe',
        action='store_true',
        help='Avoid spawning vehicles prone to accidents')
    argparser.add_argument(
        '--filterv',
        metavar='PATTERN',
        default='vehicle.*',
        help='Filter vehicle model (default: "vehicle.*")')
    argparser.add_argument(
        '--generationv',
        metavar='G',
        default='All',
        help='restrict to certain vehicle generation (values: "1","2","All" - default: "All")')
    argparser.add_argument(
        '--tm-port',
        metavar='P',
        default=8000,
        type=int,
        help='Port to communicate with TM (default: 8000)')
    argparser.add_argument(
        '--asynch',
        action='store_true',
        help='Activate asynchronous mode execution')
    argparser.add_argument(
        '--hybrid',
        action='store_true',
        help='Activate hybrid mode for Traffic Manager')
    argparser.add_argument(
        '-s', '--seed',
        metavar='S',
        type=int,
        help='Set random device seed and deterministic mode for Traffic Manager')
    argparser.add_argument(
        '--car-lights-on',
        action='store_true',
        default=False,
        help='Enable automatic car light management')
    argparser.add_argument(
        '--respawn',
        action='store_true',
        default=False,
        help='Automatically respawn dormant vehicles (only in large maps)')
    argparser.add_argument(
        '--no-rendering',
        action='store_true',
        default=False,
        help='Activate no rendering mode')

    args = argparser.parse_args()

    logging.basicConfig(format='%(levelname)s: %(message)s', level=logging.INFO)

    vehicles_list = []
    
    all_id = []
    client = carla.Client(args.host, args.port)
    client.set_timeout(10.0)
    synchronous_master = False
    random.seed(args.seed if args.seed is not None else int(time.time()))

    try:
        world = client.get_world()

        traffic_manager = client.get_trafficmanager(args.tm_port)
        traffic_manager.set_global_distance_to_leading_vehicle(2.5)
        if args.respawn:
            traffic_manager.set_respawn_dormant_vehicles(True)
        if args.hybrid:
            traffic_manager.set_hybrid_physics_mode(True)
            traffic_manager.set_hybrid_physics_radius(70.0)
        if args.seed is not None:
            traffic_manager.set_random_device_seed(args.seed)

        settings = world.get_settings()
        if not args.asynch:
            traffic_manager.set_synchronous_mode(True)
            if not settings.synchronous_mode:
                synchronous_master = True
                settings.synchronous_mode = True
                settings.fixed_delta_seconds = 0.03
            else:
                synchronous_master = False
        else:
            print("You are currently in asynchronous mode. If this is a traffic simulation, \
            you could experience some issues. If it's not working correctly, switch to synchronous \
            mode by using traffic_manager.set_synchronous_mode(True)")

        if args.no_rendering:
            settings.no_rendering_mode = True
        world.apply_settings(settings)

        blueprints = get_actor_blueprints(world, args.filterv, args.generationv)
        
        blueprints = [blueprint for blueprint in blueprints if "dreyevr" not in blueprint.id]

        if args.safe:
            blueprints = [x for x in blueprints if int(x.get_attribute('number_of_wheels')) == 4]
            blueprints = [x for x in blueprints if not x.id.endswith('microlino')]
            blueprints = [x for x in blueprints if not x.id.endswith('carlacola')]
            blueprints = [x for x in blueprints if not x.id.endswith('cybertruck')]
            blueprints = [x for x in blueprints if not x.id.endswith('t2')]
            blueprints = [x for x in blueprints if not x.id.endswith('sprinter')]
            blueprints = [x for x in blueprints if not x.id.endswith('firetruck')]
            blueprints = [x for x in blueprints if not x.id.endswith('ambulance')]

        blueprints = sorted(blueprints, key=lambda bp: bp.id)

        spawn_points = world.get_map().get_spawn_points()
        number_of_spawn_points = len(spawn_points)
        # Route 1
        spawn_point_1 =  spawn_points[20]
        # Create route 1 from the chosen spawn points
        route_1_indices = [23,1,4,5,9,10,16,17,18]
        route_1 = []
        for ind in route_1_indices:
            route_1.append(spawn_points[ind].location)

        # Route 2
        spawn_point_2 =  spawn_points[21]
        # Create route 2 from the chosen spawn points
        route_2_indices = [1,4,10,16,19]
        route_2 = []
        for ind in route_2_indices:
            route_2.append(spawn_points[ind].location)
        spawn_point_3 =  spawn_points[12]
        # Create route 1 from the chosen spawn points
        route_3_indices = [15,11,7,2,6]
        route_3 = []
        for ind in route_3_indices:
            route_3.append(spawn_points[ind].location)

        # Route 2
        spawn_point_4 =  spawn_points[0]
        # Create route 2 from the chosen spawn points
        route_4_indices = [14,13,8,3,22]
        route_4 = []
        for ind in route_4_indices:
            route_4.append(spawn_points[ind].location)
        spawn_delay = 100
        counter = spawn_delay
        alt = False

        # @todo cannot import these directly.
        SpawnActor = carla.command.SpawnActor
        SetAutopilot = carla.command.SetAutopilot
        FutureActor = carla.command.FutureActor

#         # --------------
#         # Spawn vehicles
#         # --------------
        batch = []
        n_vehicles = len(blueprints)
        while True:
            if not args.asynch and synchronous_master:
                world.tick()
                # Spawn vehicle only after delay
                
                if counter == spawn_delay and  n_vehicles < args.number_of_vehicles and n_vehicles > 0:
                    
                    vehicle_bp = random.choice(blueprints)
                    print(vehicle_bp)
                    if vehicle_bp.has_attribute("driver_id") and "dreyevr" not in vehicle_bp.id:
                        driver_id = random.choice(vehicle_bp.get_attribute('driver_id').recommended_values)

                        vehicle_bp.set_attribute('driver_id', driver_id)
                    # Alternate spawn points
                    if alt:
                        # spawn the cars and set their autopilot and light state all together
                        batch.append(SpawnActor(vehicle_bp, spawn_point_1)
                            .then(SetAutopilot(FutureActor, True, traffic_manager.get_port())))
                        #batch.append(SpawnActor(vehicle_bp, spawn_point_3).then(SetAutopilot(FutureActor, True, traffic_manager.get_port())))
                        alt = False

                    else:
                        batch.append(SpawnActor(vehicle_bp, spawn_point_2)
                            .then(SetAutopilot(FutureActor, True, traffic_manager.get_port())))
                        #batch.append(SpawnActor(vehicle_bp, spawn_point_4).then(SetAutopilot(FutureActor, True, traffic_manager.get_port())))
                        alt = True

                    vehicle = None
                    n_vehicles -= 1
                    counter -= 1
                    

                elif counter == spawn_delay and n_vehicles == 0:
                    print("Done4")
                    for response in client.apply_batch_sync(batch, synchronous_master):
                        if response.error:
                            logging.error(response.error)
                        else:
                            vehicles_list.append(response.actor_id)
                        
                        
                        all_vehicle_actors = world.get_actors(vehicles_list)
                        for vehicle in all_vehicle_actors:
                            traffic_manager.update_vehicle_lights(vehicle, True)
                    # Set parameters of TM vehicle control, we don't want lane changes
                            traffic_manager.update_vehicle_lights(vehicle, True)
                            traffic_manager.random_left_lanechange_percentage(vehicle, 0)
                            traffic_manager.random_right_lanechange_percentage(vehicle, 0)
                            traffic_manager.auto_lane_change(vehicle, False)
                            
                        # Alternate between routes
                            if alt:
                                traffic_manager.set_path(vehicle, route_1)
                                #traffic_manager.set_path(vehicle, route_3)
                                
                                alt = False
                            else:
                                traffic_manager.set_path(vehicle, route_2)
                                #traffic_manager.set_path(vehicle, route_4)
                                alt = True

                            vehicle = None

                    counter -= 1

                elif counter > 0:
                    counter -= 1
                elif counter == 0:
                    counter = spawn_delay
            else:
                world.wait_for_tick()

    finally:
        
        if not args.asynch and synchronous_master:
            settings = world.get_settings()
            settings.synchronous_mode = False
            settings.no_rendering_mode = False
            settings.fixed_delta_seconds = None
            world.apply_settings(settings)
            client.apply_batch([carla.command.DestroyActor(x) for x in vehicles_list])

        	

if __name__ == '__main__':

    try:
        main()
    except KeyboardInterrupt:
        pass
    finally:
        print('\ndone.')

**Comment by GustavoSilvera** [comment]
Hi there, are you passing the `--safe` flag in to enable `args.safe` that would enable your filter that `number_of_wheels == 4`?

**Comment by elnazbanani** [comment]
thank you, I have another issue please help me:
I search a lot and I understand try_spawn_actor could be None but how can I fixt it. my spawn point has z=0.
![image](https://user-images.githubusercontent.com/118831298/229314038-d39697ce-29d1-4b39-9f1d-fd6a9f99679f.png)
as you can see, I have vehicle_bp and also spawn point but I could noy understand why I use try I got vehicle=None 
![image](https://user-images.githubusercontent.com/118831298/229314048-b14db2eb-c9f1-4b5b-9ea1-7a63566f8121.png)
how can I fix it?

**Comment by GustavoSilvera** [comment]
Try setting the height a little higher, like z=0.1. The actor might be spawning inside the ground or your ground may not be at 0 because the map is not necessarily level everywhere. You should instead look into carla's recommended spawn-point locations to spawn your actors. 

Also you should look through [Carla's PythonAPI examples](https://github.com/carla-simulator/carla/tree/master/PythonAPI/examples) to see how they do it.

**Comment by elnazbanani** [comment]
Hi I have issues, Could you help me?
1) I want to spawn a vehicle in 4 different lanes with consideration of each route. but when I run the following code it just provided vehicles in route 1 and 3. what should I do to fix it?
2) Also, the type of vehicles was the same in each lane and did not change.
3) Although, I do not allow the vehicle to change the lanes, they changed their lane.

![image](https://user-images.githubusercontent.com/118831298/229632309-33f249da-e899-486d-bda6-e30a7d50f85f.png)


---------------------------------------code--------------------------------
import time
import carla
#import numpy as np
#from DReyeVR_utils import find_ego_vehicle
#------------------------------------------------------ 
##with VehicleLightState we can manage the position of the light in the carla:
#Class that recaps the state of the lights of a vehicle, these can be used as a flags.
#E.g: VehicleLightState.HighBeam & VehicleLightState.Brake will return True when both are active.
#Lights are off by default in any situation and should be managed by the user via script. 
#The blinkers blink automatically. Warning: Right now, not all vehicles have been prepared to
#work with this functionality, this will be added to all of them in later updates.
#------------------------------------------------------
from carla import VehicleLightState as vls

import argparse
import logging
#--------------------------------------------------------
#Logs provide developers with an extra set of eyes that are constantly looking at the flow 
#that an application is going through. They can store information, like which user or IP accessed
#the application. If an error occurs, then they can provide more insights than a stack trace by
#telling you what the state of the program was before it arrived at the line of code where the 
#error occurred.
#By logging useful data from the right places, you can not only debug errors easily but also use
#the data to analyze the performance of the application to plan for scaling or look at usage 
#patterns to plan for marketing.
#levels:debug, info, warning, error, critical
#logging.debug('This is a debug message')
#logging.info('This is an info message')
#logging.warning('This is a warning message')
#logging.error('This is an error message')
#logging.critical('This is a critical message')
#---------------------------------------------------------
from numpy import random

#---------------get actor----------------------------
def get_actor_blueprints(world, filter, generation):
    bps = world.get_blueprint_library().filter(filter)

    if generation.lower() == "all":
        return bps

    # If the filter returns only one bp, we assume that this one needed
    # and therefore, we ignore the generation
    if len(bps) == 1:
        return bps

    try:
        int_generation = int(generation)
        # Check if generation is in available generations
        if int_generation in [1, 2]:
            bps = [x for x in bps if int(x.get_attribute('generation')) == int_generation]
            return bps
        else:
            print("   Warning! Actor Generation is not valid. No actor will be spawned.")
            return []
    except:
        print("   Warning! Actor Generation is not valid. No actor will be spawned.")
        return []
#---------------------------------------------------------------
def main():
    argparser = argparse.ArgumentParser(
        description=__doc__)
    argparser.add_argument(
        '--host',
        metavar='H',
        default='127.0.0.1',
        help='IP of the host server (default: 127.0.0.1)')
    argparser.add_argument(
        '-p', '--port',
        metavar='P',
        default=2000,
        type=int,
        help='TCP port to listen to (default: 2000)')
    argparser.add_argument(
        '-n', '--number-of-vehicles',
        metavar='N',
        default=50,
        type=int,
        help='Number of vehicles (default: 30)')
    argparser.add_argument(
        '--safe',
        action='store_true',
        default= True,
        help='Avoid spawning vehicles prone to accidents')
    argparser.add_argument(
        '--filterv',
        metavar='PATTERN',
        default='vehicle.*',
        help='Filter vehicle model (default: "vehicle.*")')
    argparser.add_argument(
        '--generationv',
        metavar='G',
        default='All',
        help='restrict to certain vehicle generation (values: "1","2","All" - default: "All")')
    argparser.add_argument(
        '--tm-port',
        metavar='P',
        default=8000,
        type=int,
        help='Port to communicate with TM (default: 8000)')
    argparser.add_argument(
        '--asynch',
        action='store_true',
        help='Activate asynchronous mode execution')
    argparser.add_argument(
        '--hybrid',
        action='store_true',
        help='Activate hybrid mode for Traffic Manager')
    argparser.add_argument(
        '-s', '--seed',
        metavar='S',
        type=int,
        help='Set random device seed and deterministic mode for Traffic Manager')
    argparser.add_argument(
        '--car-lights-on',
        action='store_true',
        default=False,
        help='Enable automatic car light management')
    argparser.add_argument(
        '--respawn',
        action='store_true',
        default=False,
        help='Automatically respawn dormant vehicles (only in large maps)')
    argparser.add_argument(
        '--no-rendering',
        action='store_true',
        default=False,
        help='Activate no rendering mode')

    args = argparser.parse_args()

    logging.basicConfig(format='%(levelname)s: %(message)s', level=logging.INFO)

    vehicles_list1 = []
    vehicles_list2 = []
    vehicles_list3 = []
    vehicles_list4 = []
    all_id = []
    client = carla.Client(args.host, args.port)
    client.set_timeout(10.0)
    synchronous_master = False
    random.seed(args.seed if args.seed is not None else int(time.time()))

    try:
        world = client.get_world()

        traffic_manager = client.get_trafficmanager(args.tm_port)
        traffic_manager.set_global_distance_to_leading_vehicle(2.5)
        if args.respawn:
            traffic_manager.set_respawn_dormant_vehicles(True)
        if args.hybrid:
            traffic_manager.set_hybrid_physics_mode(True)
            traffic_manager.set_hybrid_physics_radius(70.0)
        if args.seed is not None:
            traffic_manager.set_random_device_seed(args.seed)

        settings = world.get_settings()
        if not args.asynch:
            traffic_manager.set_synchronous_mode(True)
            if not settings.synchronous_mode:
                synchronous_master = True
                settings.synchronous_mode = True
                settings.fixed_delta_seconds = 0.03
            else:
                synchronous_master = False
        else:
            print("You are currently in asynchronous mode. If this is a traffic simulation, \
            you could experience some issues. If it's not working correctly, switch to synchronous \
            mode by using traffic_manager.set_synchronous_mode(True)")

        if args.no_rendering:
            settings.no_rendering_mode = True
        world.apply_settings(settings)

        blueprints = get_actor_blueprints(world, args.filterv, args.generationv)
        blueprints = [blueprint for blueprint in blueprints if "dreyevr" not in blueprint.id]


        if args.safe:
            blueprints = [x for x in blueprints if int(x.get_attribute('number_of_wheels')) == 4]
            blueprints = [x for x in blueprints if not x.id.endswith('microlino')]
            blueprints = [x for x in blueprints if not x.id.endswith('carlacola')]
            blueprints = [x for x in blueprints if not x.id.endswith('cybertruck')]
            blueprints = [x for x in blueprints if not x.id.endswith('t2')]
            blueprints = [x for x in blueprints if not x.id.endswith('sprinter')]
            blueprints = [x for x in blueprints if not x.id.endswith('firetruck')]
            blueprints = [x for x in blueprints if not x.id.endswith('ambulance')]

        blueprints = sorted(blueprints, key=lambda bp: bp.id)
        
        spawn_points = world.get_map().get_spawn_points()
        number_of_spawn_points = len(spawn_points)
        # Route 1
        spawn_point_1 =  spawn_points[0]
        # Create route 1 from the chosen spawn points
        route_1_indices = [15,1,2,3,4,5,6,7,8]
        route_1 = []
        for ind in route_1_indices:
            route_1.append(spawn_points[ind].location)

        # Route 2
        spawn_point_2 =  spawn_points[11]
        # Create route 2 from the chosen spawn points
        route_2_indices = [1,2,5,6,9]
        route_2 = []
        for ind in route_2_indices:
            route_2.append(spawn_points[ind].location)
        # Route 3
        spawn_point_3 =  spawn_points[10]
        # Create route 3 from the chosen spawn points
        route_3_indices = [13]
        route_3 = []
        for ind in route_3_indices:
            route_3.append(spawn_points[ind].location)

        # Route 4
        spawn_point_4 =  spawn_points[12]
        # Create route 4 from the chosen spawn points
        route_4_indices = [14]
        route_4 = []
        for ind in route_4_indices:
            route_4.append(spawn_points[ind].location)
        spawn_delay = 200
        counter = spawn_delay
        alt = True
        alt2 = True

        # @todo cannot import these directly.
        SpawnActor = carla.command.SpawnActor
        SetAutopilot = carla.command.SetAutopilot
        FutureActor = carla.command.FutureActor

#         # --------------
#         # Spawn vehicles
#         # --------------
        batch1 = []
        batch2 = []
        batch3 = []
        batch4 = []
        n_vehicles = len(blueprints)
        while True:
            if not args.asynch and synchronous_master:
                world.tick()
                # Spawn vehicle only after delay
                vehicle_bp = random.choice(blueprints)
                if vehicle_bp.has_attribute("driver_id") and "dreyevr" not in vehicle_bp.id:
                    driver_id = random.choice(vehicle_bp.get_attribute('driver_id').recommended_values)
                    vehicle_bp.set_attribute('driver_id', driver_id)
                if counter == spawn_delay and  n_vehicles < args.number_of_vehicles and n_vehicles > 0:
                    
                    # Alternate spawn points
                    if alt:
                        if alt2:
                            batch1.append(SpawnActor(vehicle_bp, spawn_point_1).then(SetAutopilot(FutureActor, True, traffic_manager.get_port())))
                        else:
                            batch3.append(SpawnActor(vehicle_bp, spawn_point_3).then(SetAutopilot(FutureActor, True, traffic_manager.get_port())))
                            
                            
                    else:
                        if alt2:
                            batch2.append(SpawnActor(vehicle_bp, spawn_point_2).then(SetAutopilot(FutureActor, True, traffic_manager.get_port())))
                           
                            
                        else:
                            batch4.append(SpawnActor(vehicle_bp, spawn_point_4).then(SetAutopilot(FutureActor, True, traffic_manager.get_port())))
                            
                            

                    for response1 in client.apply_batch_sync(batch1, synchronous_master):
                        if response1.error:
                            logging.error(response1.error)
                        else:
                            vehicles_list1.append(response1.actor_id)   

                    for response2 in client.apply_batch_sync(batch2, synchronous_master):
                        if response2.error:
                            logging.error(response2.error)
                        else:
                            vehicles_list2.append(response2.actor_id)

                    for response3 in client.apply_batch_sync(batch3, synchronous_master):
                        if response3.error:
                            logging.error(response3.error)
                        else:
                            vehicles_list3.append(response3.actor_id)

                    for response4 in client.apply_batch_sync(batch4, synchronous_master):
                        if response4.error:
                            logging.error(response4.error)
                        else:
                            vehicles_list4.append(response4.actor_id)
                    

                
                            
                        
                            
                            
                    # Alternate between routes
                    if alt:
                        if alt2:
                            all_vehicle_actors1 = world.get_actors(vehicles_list1)
                            for vehicle1 in all_vehicle_actors1:
                                traffic_manager.update_vehicle_lights(vehicle1, True)
                                traffic_manager.random_left_lanechange_percentage(vehicle1, 0)
                                traffic_manager.random_right_lanechange_percentage(vehicle1, 0)
                                traffic_manager.auto_lane_change(vehicle1, False)
                                traffic_manager.set_path(vehicle1, route_1)
                                print("1")
                            alt = True
                            alt2 = False
                        else:
                            all_vehicle_actors3 = world.get_actors(vehicles_list3)
                            for vehicle3 in all_vehicle_actors3:
                                traffic_manager.update_vehicle_lights(vehicle3, True)
                                traffic_manager.random_left_lanechange_percentage(vehicle3, 0)
                                traffic_manager.random_right_lanechange_percentage(vehicle3, 0)
                                traffic_manager.auto_lane_change(vehicle3, False)
                                traffic_manager.set_path(vehicle3, route_3)
                                print("3")
                            alt = False
                            alt2 = True
                    else:
                        if alt2:
                            all_vehicle_actors2 = world.get_actors(vehicles_list2)
                            for vehicle2 in all_vehicle_actors2:
                                traffic_manager.update_vehicle_lights(vehicle2, True)
                                traffic_manager.random_left_lanechange_percentage(vehicle2, 0)
                                traffic_manager.random_right_lanechange_percentage(vehicle2, 0)
                                traffic_manager.auto_lane_change(vehicle2, False)
                                traffic_manager.set_path(vehicle2, route_2)
                                print("2")
                            alt = False
                            alt2 = False
                        else:
                            all_vehicle_actors4 = world.get_actors(vehicles_list4)
                            for vehicle4 in all_vehicle_actors4:
                                traffic_manager.update_vehicle_lights(vehicle4, True)
                                traffic_manager.random_left_lanechange_percentage(vehicle4, 0)
                                traffic_manager.random_right_lanechange_percentage(vehicle4, 0)
                                traffic_manager.auto_lane_change(vehicle4, False)
                                traffic_manager.set_path(vehicle4, route_4)
                                print("4")
                            
                            alt = True
                            alt2 = True

                    

                    counter -= 1

                elif counter > 0:
                    counter -= 1
                elif counter == 0:
                    counter = spawn_delay
            else:
                world.wait_for_tick()




    finally:
        
        if not args.asynch and synchronous_master:
            settings = world.get_settings()
            settings.synchronous_mode = False
            settings.no_rendering_mode = False
            settings.fixed_delta_seconds = None
            world.apply_settings(settings)
            client.apply_batch([carla.command.DestroyActor(x) for x in vehicles_list1])
            client.apply_batch([carla.command.DestroyActor(x) for x in vehicles_list2])
            client.apply_batch([carla.command.DestroyActor(x) for x in vehicles_list3])
            client.apply_batch([carla.command.DestroyActor(x) for x in vehicles_list4])

        	

if __name__ == '__main__':

    try:
        main()
    except KeyboardInterrupt:
        pass
    finally:
        print('\ndone.')

**Comment by elnazbanani** [comment]
> Try setting the height a little higher, like z=0.1. The actor might be spawning inside the ground or your ground may not be at 0 because the map is not necessarily level everywhere. You should instead look into carla's recommended spawn-point locations to spawn your actors.
> 
> Also you should look through [Carla's PythonAPI examples](https://github.com/carla-simulator/carla/tree/master/PythonAPI/examples) to see how they do it.

**Comment by elnazbanani** [comment]
Hi, I want to generate vehicles in 4 different lanes but I have some issues as you can see in the following picture:
1) vehicles just provided in lane 1 and 3.
2)all of the vehicles are the same.
3)Although I did not allow them to change their lane, they changed it.
Could you help me?
![image](https://user-images.githubusercontent.com/118831298/229633910-a57d8bf4-4a58-41d4-a1d4-8ec1e93bf6b7.png)
---------------------code------------------------
import time
import carla
from carla import VehicleLightState as vls
import argparse
import logging
from numpy import random
def get_actor_blueprints(world, filter, generation):
    bps = world.get_blueprint_library().filter(filter)

    if generation.lower() == "all":
        return bps

    # If the filter returns only one bp, we assume that this one needed
    # and therefore, we ignore the generation
    if len(bps) == 1:
        return bps

    try:
        int_generation = int(generation)
        # Check if generation is in available generations
        if int_generation in [1, 2]:
            bps = [x for x in bps if int(x.get_attribute('generation')) == int_generation]
            return bps
        else:
            print("   Warning! Actor Generation is not valid. No actor will be spawned.")
            return []
    except:
        print("   Warning! Actor Generation is not valid. No actor will be spawned.")
        return []
#---------------------------------------------------------------
def main():
    argparser = argparse.ArgumentParser(
        description=__doc__)
    argparser.add_argument(
        '--host',
        metavar='H',
        default='127.0.0.1',
        help='IP of the host server (default: 127.0.0.1)')
    argparser.add_argument(
        '-p', '--port',
        metavar='P',
        default=2000,
        type=int,
        help='TCP port to listen to (default: 2000)')
    argparser.add_argument(
        '-n', '--number-of-vehicles',
        metavar='N',
        default=50,
        type=int,
        help='Number of vehicles (default: 30)')
    argparser.add_argument(
        '--safe',
        action='store_true',
        default= True,
        help='Avoid spawning vehicles prone to accidents')
    argparser.add_argument(
        '--filterv',
        metavar='PATTERN',
        default='vehicle.*',
        help='Filter vehicle model (default: "vehicle.*")')
    argparser.add_argument(
        '--generationv',
        metavar='G',
        default='All',
        help='restrict to certain vehicle generation (values: "1","2","All" - default: "All")')
    argparser.add_argument(
        '--tm-port',
        metavar='P',
        default=8000,
        type=int,
        help='Port to communicate with TM (default: 8000)')
    argparser.add_argument(
        '--asynch',
        action='store_true',
        help='Activate asynchronous mode execution')
    argparser.add_argument(
        '--hybrid',
        action='store_true',
        help='Activate hybrid mode for Traffic Manager')
    argparser.add_argument(
        '-s', '--seed',
        metavar='S',
        type=int,
        help='Set random device seed and deterministic mode for Traffic Manager')
    argparser.add_argument(
        '--car-lights-on',
        action='store_true',
        default=False,
        help='Enable automatic car light management')
    argparser.add_argument(
        '--respawn',
        action='store_true',
        default=False,
        help='Automatically respawn dormant vehicles (only in large maps)')
    argparser.add_argument(
        '--no-rendering',
        action='store_true',
        default=False,
        help='Activate no rendering mode')

    args = argparser.parse_args()

    logging.basicConfig(format='%(levelname)s: %(message)s', level=logging.INFO)

    vehicles_list1 = []
    vehicles_list2 = []
    vehicles_list3 = []
    vehicles_list4 = []
    all_id = []
    client = carla.Client(args.host, args.port)
    client.set_timeout(10.0)
    synchronous_master = False
    random.seed(args.seed if args.seed is not None else int(time.time()))

    try:
        world = client.get_world()

        traffic_manager = client.get_trafficmanager(args.tm_port)
        traffic_manager.set_global_distance_to_leading_vehicle(2.5)
        if args.respawn:
            traffic_manager.set_respawn_dormant_vehicles(True)
        if args.hybrid:
            traffic_manager.set_hybrid_physics_mode(True)
            traffic_manager.set_hybrid_physics_radius(70.0)
        if args.seed is not None:
            traffic_manager.set_random_device_seed(args.seed)

        settings = world.get_settings()
        if not args.asynch:
            traffic_manager.set_synchronous_mode(True)
            if not settings.synchronous_mode:
                synchronous_master = True
                settings.synchronous_mode = True
                settings.fixed_delta_seconds = 0.03
            else:
                synchronous_master = False
        else:
            print("You are currently in asynchronous mode. If this is a traffic simulation, \
            you could experience some issues. If it's not working correctly, switch to synchronous \
            mode by using traffic_manager.set_synchronous_mode(True)")

        if args.no_rendering:
            settings.no_rendering_mode = True
        world.apply_settings(settings)

        blueprints = get_actor_blueprints(world, args.filterv, args.generationv)
        blueprints = [blueprint for blueprint in blueprints if "dreyevr" not in blueprint.id]


        if args.safe:
            blueprints = [x for x in blueprints if int(x.get_attribute('number_of_wheels')) == 4]
            blueprints = [x for x in blueprints if not x.id.endswith('microlino')]
            blueprints = [x for x in blueprints if not x.id.endswith('carlacola')]
            blueprints = [x for x in blueprints if not x.id.endswith('cybertruck')]
            blueprints = [x for x in blueprints if not x.id.endswith('t2')]
            blueprints = [x for x in blueprints if not x.id.endswith('sprinter')]
            blueprints = [x for x in blueprints if not x.id.endswith('firetruck')]
            blueprints = [x for x in blueprints if not x.id.endswith('ambulance')]

        blueprints = sorted(blueprints, key=lambda bp: bp.id)
        
        spawn_points = world.get_map().get_spawn_points()
        number_of_spawn_points = len(spawn_points)
        # Route 1
        spawn_point_1 =  spawn_points[0]
        # Create route 1 from the chosen spawn points
        route_1_indices = [15,1,2,3,4,5,6,7,8]
        route_1 = []
        for ind in route_1_indices:
            route_1.append(spawn_points[ind].location)

        # Route 2
        spawn_point_2 =  spawn_points[11]
        # Create route 2 from the chosen spawn points
        route_2_indices = [1,2,5,6,9]
        route_2 = []
        for ind in route_2_indices:
            route_2.append(spawn_points[ind].location)
        # Route 3
        spawn_point_3 =  spawn_points[10]
        # Create route 3 from the chosen spawn points
        route_3_indices = [13]
        route_3 = []
        for ind in route_3_indices:
            route_3.append(spawn_points[ind].location)

        # Route 4
        spawn_point_4 =  spawn_points[12]
        # Create route 4 from the chosen spawn points
        route_4_indices = [14]
        route_4 = []
        for ind in route_4_indices:
            route_4.append(spawn_points[ind].location)
        spawn_delay = 200
        counter = spawn_delay
        alt = True
        alt2 = True

        # @todo cannot import these directly.
        SpawnActor = carla.command.SpawnActor
        SetAutopilot = carla.command.SetAutopilot
        FutureActor = carla.command.FutureActor
        batch1 = []
        batch2 = []
        batch3 = []
        batch4 = []
        n_vehicles = len(blueprints)
        while True:
            if not args.asynch and synchronous_master:
                world.tick()
                # Spawn vehicle only after delay
                vehicle_bp = random.choice(blueprints)
                if vehicle_bp.has_attribute("driver_id") and "dreyevr" not in vehicle_bp.id:
                    driver_id = random.choice(vehicle_bp.get_attribute('driver_id').recommended_values)
                    vehicle_bp.set_attribute('driver_id', driver_id)
                if counter == spawn_delay and  n_vehicles < args.number_of_vehicles and n_vehicles > 0:
                    
                    # Alternate spawn points
                    if alt:
                        if alt2:
                            batch1.append(SpawnActor(vehicle_bp, spawn_point_1).then(SetAutopilot(FutureActor, True, traffic_manager.get_port())))
                        else:
                            batch3.append(SpawnActor(vehicle_bp, spawn_point_3).then(SetAutopilot(FutureActor, True, traffic_manager.get_port())))
                            
                            
                    else:
                        if alt2:
                            batch2.append(SpawnActor(vehicle_bp, spawn_point_2).then(SetAutopilot(FutureActor, True, traffic_manager.get_port())))
                           
                            
                        else:
                            batch4.append(SpawnActor(vehicle_bp, spawn_point_4).then(SetAutopilot(FutureActor, True, traffic_manager.get_port())))
                            
                            

                    for response1 in client.apply_batch_sync(batch1, synchronous_master):
                        if response1.error:
                            logging.error(response1.error)
                        else:
                            vehicles_list1.append(response1.actor_id)   

                    for response2 in client.apply_batch_sync(batch2, synchronous_master):
                        if response2.error:
                            logging.error(response2.error)
                        else:
                            vehicles_list2.append(response2.actor_id)

                    for response3 in client.apply_batch_sync(batch3, synchronous_master):
                        if response3.error:
                            logging.error(response3.error)
                        else:
                            vehicles_list3.append(response3.actor_id)

                    for response4 in client.apply_batch_sync(batch4, synchronous_master):
                        if response4.error:
                            logging.error(response4.error)
                        else:
                            vehicles_list4.append(response4.actor_id)
                    

                
                            
                        
                            
                            
                    # Alternate between routes
                    if alt:
                        if alt2:
                            all_vehicle_actors1 = world.get_actors(vehicles_list1)
                            for vehicle1 in all_vehicle_actors1:
                                traffic_manager.update_vehicle_lights(vehicle1, True)
                                traffic_manager.random_left_lanechange_percentage(vehicle1, 0)
                                traffic_manager.random_right_lanechange_percentage(vehicle1, 0)
                                traffic_manager.auto_lane_change(vehicle1, False)
                                traffic_manager.set_path(vehicle1, route_1)
                                print("1")
                            alt = True
                            alt2 = False
                        else:
                            all_vehicle_actors3 = world.get_actors(vehicles_list3)
                            for vehicle3 in all_vehicle_actors3:
                                traffic_manager.update_vehicle_lights(vehicle3, True)
                                traffic_manager.random_left_lanechange_percentage(vehicle3, 0)
                                traffic_manager.random_right_lanechange_percentage(vehicle3, 0)
                                traffic_manager.auto_lane_change(vehicle3, False)
                                traffic_manager.set_path(vehicle3, route_3)
                                print("3")
                            alt = False
                            alt2 = True
                    else:
                        if alt2:
                            all_vehicle_actors2 = world.get_actors(vehicles_list2)
                            for vehicle2 in all_vehicle_actors2:
                                traffic_manager.update_vehicle_lights(vehicle2, True)
                                traffic_manager.random_left_lanechange_percentage(vehicle2, 0)
                                traffic_manager.random_right_lanechange_percentage(vehicle2, 0)
                                traffic_manager.auto_lane_change(vehicle2, False)
                                traffic_manager.set_path(vehicle2, route_2)
                                print("2")
                            alt = False
                            alt2 = False
                        else:
                            all_vehicle_actors4 = world.get_actors(vehicles_list4)
                            for vehicle4 in all_vehicle_actors4:
                                traffic_manager.update_vehicle_lights(vehicle4, True)
                                traffic_manager.random_left_lanechange_percentage(vehicle4, 0)
                                traffic_manager.random_right_lanechange_percentage(vehicle4, 0)
                                traffic_manager.auto_lane_change(vehicle4, False)
                                traffic_manager.set_path(vehicle4, route_4)
                                print("4")
                            
                            alt = True
                            alt2 = True

                    

                    counter -= 1

                elif counter > 0:
                    counter -= 1
                elif counter == 0:
                    counter = spawn_delay
            else:
                world.wait_for_tick()
    finally:
        
        if not args.asynch and synchronous_master:
            settings = world.get_settings()
            settings.synchronous_mode = False
            settings.no_rendering_mode = False
            settings.fixed_delta_seconds = None
            world.apply_settings(settings)
            client.apply_batch([carla.command.DestroyActor(x) for x in vehicles_list1])
            client.apply_batch([carla.command.DestroyActor(x) for x in vehicles_list2])
            client.apply_batch([carla.command.DestroyActor(x) for x in vehicles_list3])
            client.apply_batch([carla.command.DestroyActor(x) for x in vehicles_list4])
